File: 0701-0800/791-800/791M_Custom_Sort_String/cpp/sorting.cpp
// 791. Custom Sort String
// https://leetcode.com/problems/custom-sort-string/description/
// T.C.: O(n log n)
// S.C.: O(n)
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>
using namespace std;

class Solution
{
public:
  string customSortString(string order, string s)
  {
    sort(begin(s), end(s), [&](char a, char b)
         { return order.find(a) < order.find(b); });
    return s;
  }
};

// g++ -std=c++2a app.cpp -o ./app.exe && ./app
int main()
{
  string order = "cba";
  string s = "abcd";
  string expected = "cbad";
  string result = Solution().customSortString(order, s);
  cout << result << " " << (expected == result ? "true" : "false") << endl;

  order = "bcafg";
  s = "abcd";
  expected = "bcad";
  result = Solution().customSortString(order, s);
  cout << result << " " << (expected == result ? "true" : "false") << endl;

  return 0;
}

File: 0701-0800/791-800/791M_Custom_Sort_String/sorting.js
// 791. Custom Sort String
// https://leetcode.com/problems/custom-sort-string/description/
// T.C.: O(n log n)
// S.C.: O(n)
/**
 * @param {string} order
 * @param {string} s
 * @return {string}
 */
var customSortString = function (order, s) {
  const map = new Map();
  for (let i = 0; i < order.length; i++) {
    map.set(order[i], i);
  }
  const result = [...s];
  result.sort((a, b) => {
    if (map.has(a) && map.has(b)) {
      return map.get(a) - map.get(b);
    } else if (map.has(a)) {
      return -1;
    } else if (map.has(b)) {
      return 1;
    } else {
      return 0;
    }
  });
  return result.join('');
};

var order = 'cba',
  s = 'abcd';
var expected = 'cbad';
var result = customSortString(order, s);
console.log(result, result === expected);

var order = 'bcafg',
  s = 'abcd';
var expected = 'bcad';
var result = customSortString(order, s);
console.log(result, result === expected);
